testing/separated/places-bench/src/database.rs

```diff
@@ -130,3 +130,284 @@ pub fn bench_match_url(c: &mut Criterion) {
+    db.execute(
+        "CREATE INDEX IF NOT EXISTS idx_places_outgoing_by_frecency
+         ON moz_places(frecency DESC)
+         WHERE hidden = 0 AND (sync_change_counter > 0 OR sync_status != 2)",
+        [],
+    )?;
+    db.execute("ANALYZE", [])?;
+    Ok(())
+}
+
+fn run_outgoing_query(
+    db: &PlacesDb,
+    use_index_friendly_where: bool,
+    limit: usize,
+) -> places::Result<usize> {
+    let sql_tmpl = if use_index_friendly_where {
+        OUTGOING_SQL_INDEX_FRIENDLY
+    } else {
+        OUTGOING_SQL_ORIGINAL
+    };
+    let sql = sql_tmpl.replace("{LIMIT}", &limit.to_string());
+
+    let mut stmt = db.prepare_maybe_cached(&sql, true)?;
+    let mut rows = stmt.query([])?;
+    let mut count = 0usize;
+    while let Some(_row) = rows.next()? {
+        count += 1;
+    }
+    Ok(count)
+}
+
+pub fn bench_outgoing_candidates(c: &mut Criterion) {
+    // Create two independent DBs so index state changes do not leak between benches.
+    let db_no_index = TestDb::new();
+    seed_outgoing_flags(&db_no_index.db).unwrap();
+    drop_outgoing_partial_index(&db_no_index.db).unwrap();
+    explain(
+        "Original query".to_string(),
+        &db_no_index.db,
+        OUTGOING_SQL_ORIGINAL,
+        200,
+    );
+
+    let db_with_index = TestDb::new();
+    seed_outgoing_flags(&db_with_index.db).unwrap();
+    create_outgoing_partial_index(&db_with_index.db).unwrap();
+    explain(
+        "Index friendly query".to_string(),
+        &db_with_index.db,
+        OUTGOING_SQL_INDEX_FRIENDLY,
+        200,
+    );
+
+    // Bench: no index, original predicate
+    {
+        let test_db = db_no_index.clone();
+        c.bench_function(
+            "outgoing_candidates: original WHERE, NO partial index",
+            move |b| {
+                let db = &test_db.db;
+                b.iter(|| run_outgoing_query(db, /*use_index_friendly_where=*/ false, 200).unwrap())
+            },
+        );
+    }
+
+    // Bench: partial index + index-friendly predicate
+    {
+        let test_db = db_with_index.clone();
+        c.bench_function(
+            "outgoing_candidates: index-friendly WHERE + partial index",
+            move |b| {
+                let db = &test_db.db;
+                b.iter(|| run_outgoing_query(db, /*use_index_friendly_where=*/ true, 200).unwrap())
+            },
+        );
+    }
+
+    // also measure “partial index + original WHERE”.
+    {
+        let test_db = db_with_index.clone();
+        c.bench_function(
+            "outgoing_candidates: original WHERE + partial index",
+            move |b| {
+                let db = &test_db.db;
+                b.iter(|| run_outgoing_query(db, /*use_index_friendly_where=*/ false, 200).unwrap())
+            },
+        );
+    }
+}
+
+/*
+ * Benchmarking speeding up get_top_frecent_site_infos by adding a partial index to help speed up the query
+*/
+
+const VISIBLE_BY_FRECENCY_SQL: &str = r#"
+    SELECT h.frecency, h.title, h.url
```

Comment:
this isn't the exact query used. When I tried the exact query you used, I didn't see the new index - I did however see the `frecencyindex` used in both cases. Once we get this right, should we remove that other index? Regardless, it looks like this should maybe copy both the old and new exact sql like you did above?